src/main/java/net/neoforged/neoforge/client/event/AddSectionGeometryEvent.java

```diff
@@ -0,0 +1,163 @@
+     */
+    public void addRenderer(AdditionalSectionRenderer renderer) {
+        additionalRenderers.add(renderer);
+    }
+
+    /**
+     * @return the list of all added renderers. Do not modify the result.
+     */
+    public List<AdditionalSectionRenderer> getAdditionalRenderers() {
+        return additionalRenderers;
+    }
+
+    /**
+     * @return the origin of the section to add renderers to, i.e. the block with the smallest coordinates contained in
+     *         the section.
+     */
+    public BlockPos getSectionOrigin() {
+        return sectionOrigin;
+    }
+
+    /**
+     * @return the level to render in. This can differ from the current client level in case of e.g. guidebooks.
+     */
+    public Level getLevel() {
+        Preconditions.checkState(Minecraft.getInstance().isSameThread());
+        return level;
+    }
+
+    /**
+     * A rendering callback that will be invoked during chunk meshing.
+     */
+    @FunctionalInterface
+    public interface AdditionalSectionRenderer {
+        void render(SectionRenderingContext context);
+    }
+
+    public static final class SectionRenderingContext {
+        private final Function<RenderType, VertexConsumer> getOrCreateLayer;
+        private final BlockAndTintGetter region;
+        private final PoseStack poseStack;
+
+        /**
+         * @param getOrCreateLayer a function that, given a "chunk render type", returns the corresponding buffer and
+         *                         adds it to the section if it is not already present.
+         * @param region           a view of the section and some surrounding blocks
+         * @param poseStack        the transformations to use, currently set to the chunk origin at unit scaling and no
+         *                         rotation.
+         */
+        public SectionRenderingContext(
+                Function<RenderType, VertexConsumer> getOrCreateLayer, BlockAndTintGetter region, PoseStack poseStack) {
+            this.getOrCreateLayer = getOrCreateLayer;
+            this.region = region;
+            this.poseStack = poseStack;
+        }
+
+        /**
+         * Returns the builder for the given render type/layer in the chunk section. If the render type is not already
+         * present in the section, marks the type as present in the section.
+         * 
+         * @param type the render type to retrieve a builder for. This has to be one of the render types listed in
+         *             {@link net.minecraft.client.renderer.RenderType#chunkBufferLayers}.
+         * @return a vertex consumer adding geometry of the specified layer
+         * @throws IllegalArgumentException if the passed render type is not in
+         *                                  {@link net.minecraft.client.renderer.RenderType#chunkBufferLayers}.
+         */
+        public VertexConsumer getOrCreateChunkBuffer(RenderType type) {
+            Preconditions.checkArgument(RenderType.chunkBufferLayers().contains(type));
```

Comment:
`type.getChunkLayerId() >= 0` is probably slightly more efficient than scanning the list of chunk buffer layers.

Reply:
Ah, I did not know that method existed.